### DISCUSSION TITLE
为什么 TCP 协议有性能问题 · Why's THE Design? · /whys-the-design-tcp-performance

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#584] [Announcements]
https://draveness.me/whys-the-design-tcp-performance

**Comment by tamarous** [comment]
你好，文章中RTT的英文全称疑有误，维基上是Round-Trip Time

**Comment by draveness** [comment]
> @tamarous
> 你好，文章中RTT的英文全称疑有误，维基上是Round-Trip Time

打错了...我改下...

**Comment by Mylog** [comment]
性能优化的一个点在文章说拥塞窗口增长的时候就体现了，一次发送了10 or 20 个包，也就是MSS。
kafka的设计里也体现了这点

**Comment by leafduo** [comment]
对应三个问题看 QUIC 是怎么做的：

1. 拥塞控制算法目前还是用的 TCP 那一些，但是转移到用户空间之后就方便改了
2. 建立连接所需的 RTT 变少了，还有连接迁移之类的特性使得有些情况不需要再建立连接了
3. QUIC 能传好多 SACK 范围

这些改进应该并没有更科学的拥塞控制算法有用，可能还是需要期待不基于丢包的拥塞控制算法，比如 BBR

**Comment by draveness** [comment]
> 这些改进应该并没有更科学的拥塞控制算法有用，可能还是需要期待不基于丢包的拥塞控制算法，比如 BBR

说得对，**丢包意味着网络拥塞这一假设在很多时候已经不成立**了，这也是问题的根源。

**Comment by njuxjy** [comment]
关于线增积减的解释感觉不太准确：

> 线性增长：当发送方收到了接收方的 ACK 时，拥塞窗口大小会加一；

每收到一个ACK就+1的话应该还处于指数增长阶段，即`cwnd < ssthresh`时。当`cwnd > ssthresh`时，每经过一个RTT就把cwnd加1，此时对于cwnd来说是"线增"了。

> 积式减少：当发送方发送的数据包丢包时，拥塞控制窗口会减半；

发生丢包时，减半的应该是`ssthresh`，准确来说是减为当前发送方`cwnd`的一半，而`cwnd`是被置为1，此时对于`ssthresh`来说是`积减`了。

**Comment by draveness** [comment]
> 关于线增积减的解释感觉不太准确：
> 
> > 线性增长：当发送方收到了接收方的 ACK 时，拥塞窗口大小会加一；
> 
> 每收到一个ACK就+1的话应该还处于指数增长阶段，即`cwnd < ssthresh`时。当`cwnd > ssthresh`时，每经过一个RTT就把cwnd加1，此时对于cwnd来说是"线增"了。
> 
> > 积式减少：当发送方发送的数据包丢包时，拥塞控制窗口会减半；
> 
> 发生丢包时，减半的应该是`ssthresh`，准确来说是减为当前发送方`cwnd`的一半，而`cwnd`是被置为1，此时对于`ssthresh`来说是`积减`了。

你说得对，这里应该展开详细说一下，我明天改下。

**Comment by NeoTse** [comment]
文章中关于TCP握手时的数据帧大小有异议。通过固网抓包验证，TCP ACK和TCP ACK+SYN的大小为 32 bytes（其中options大小为12bytes），TCP SYN的大小为 20 bytes（无options）。[TCP header的Data offset是包含options长度的](https://en.wikipedia.org/wiki/Transmission_Control_Protocol#TCP_segment_structure)。

**Comment by draveness** [comment]
> 文章中关于TCP握手时的数据帧大小有异议。通过固网抓包验证，TCP ACK和TCP ACK+SYN的大小为 32 bytes（其中options大小为12bytes），TCP SYN的大小为 20 bytes（无options）。[TCP header的Data offset是包含options长度的](https://en.wikipedia.org/wiki/Transmission_Control_Protocol#TCP_segment_structure)。

文中关于 TCP 三次握手数据包大小也是通过抓包获取的，访问不同的服务器确实可能会造成包含 Options 的 TCP 协议头大小不一致，我会修改文中对应的段落，变的更精确。

![image](https://user-images.githubusercontent.com/6493255/71947523-19ab3800-3208-11ea-979a-21b4e95ad176.png)

三个数据包长度为 78、78 和 54，其中 14 字节为以太网首部，20 字节为 IP 首部。

**Comment by NeoTse** [comment]
> > 文章中关于TCP握手时的数据帧大小有异议。通过固网抓包验证，TCP ACK和TCP ACK+SYN的大小为 32 bytes（其中options大小为12bytes），TCP SYN的大小为 20 bytes（无options）。[TCP header的Data offset是包含options长度的](https://en.wikipedia.org/wiki/Transmission_Control_Protocol#TCP_segment_structure)。
> 
> 文中关于 TCP 三次握手数据包大小也是通过抓包获取的，访问不同的服务器确实可能会造成包含 Options 的 TCP 协议头大小不一致，我会修改文中对应的段落，变的更精确。
> 
> ![image](https://user-images.githubusercontent.com/6493255/71947523-19ab3800-3208-11ea-979a-21b4e95ad176.png)
> 
> 三个数据包长度为 78、78 和 54，其中 14 字节为以太网首部，20 字节为 IP 首部。

嗯，这个大小会随着网络环境浮动（20 <= Length <= 60）

**Comment by AK47777** [comment]
我是一个流媒体的小白，因为处理视频相关一直想问题个问题。tcp是否确实是永远都不会丢包？就是只要丢包就会一直传，这样会不会导致网络最后被塞满。我的理解是丢了再发，不断重传早晚塞满。另外这样重发会不会很慢，因为重传导致要很久才能把包拼凑起来。导致我们上层处理数据很长时间没有数据。

**Comment by draveness** [comment]
> 我是一个流媒体的小白，因为处理视频相关一直想问题个问题。tcp是否确实是永远都不会丢包？就是只要丢包就会一直传，这样会不会导致网络最后被塞满。我的理解是丢了再发，不断重传早晚塞满。另外这样重发会不会很慢，因为重传导致要很久才能把包拼凑起来。导致我们上层处理数据很长时间没有数据。

你说的对。流媒体一般都会用 UDP 吧，你们用 TCP 传数据？

**Comment by AK47777** [comment]
安防行业，国标2016里面规定了TCP，所以一定要使用TCP的。所以我还想请继续解答我的问题：1、有没有这样一种可能，在3%的丢包率下，有一个包比较倒霉，从第一次开始，就一直被丢，一直被重传。丢了50次。但是TCP协议层一定要等着收到这个包，编好顺序才能送给上层。这样就会导致后面来的数据因为前面的数据不全，就一直传不下去？导致我应用层收不到数据？

**Comment by draveness** [comment]
> 安防行业，国标2016里面规定了TCP，所以一定要使用TCP的。

竟然还有这种规定么，确实是第一回听说

> 所以我还想请继续解答我的问题：1、有没有这样一种可能，在3%的丢包率下，有一个包比较倒霉，从第一次开始，就一直被丢，一直被重传。丢了50次。但是TCP协议层一定要等着收到这个包，编好顺序才能送给上层。这样就会导致后面来的数据因为前面的数据不全，就一直传不下去？导致我应用层收不到数据？

如果接收方收到的数据是不连续的，应用层也没法识别啊，例如一个 JSON，如果你只收到了前半段和后半段，中间少了一部分数据，也拼不成整个 JSON...

**Comment by AK47777** [comment]
> @draveness
> > 安防行业，国标2016里面规定了TCP，所以一定要使用TCP的。
> 
> 竟然还有这种规定么，确实是第一回听说
> 
> > 所以我还想请继续解答我的问题：1、有没有这样一种可能，在3%的丢包率下，有一个包比较倒霉，从第一次开始，就一直被丢，一直被重传。丢了50次。但是TCP协议层一定要等着收到这个包，编好顺序才能送给上层。这样就会导致后面来的数据因为前面的数据不全，就一直传不下去？导致我应用层收不到数据？
> 
> 如果接收方收到的数据是不连续的，应用层也没法识别啊，例如一个 JSON，如果你只收到了前半段和后半段，中间少了一部分数据，也拼不成整个 JSON...

所以我一直想问的就是，在一个有3%的丢包率的网络环境下，一个包如果运气不好一直被丢。那他会一直被重传吗？还是过了几次之后就不再传了？

**Comment by draveness** [comment]
> 所以我一直想问的就是，在一个有3%的丢包率的网络环境下，一个包如果运气不好一直被丢。那他会一直被重传吗？还是过了几次之后就不再传了？

可以看一下 Linux 的手册 https://linux.die.net/man/7/tcp

> tcp_retries1 (integer; **default: 3**; since Linux 2.2)
The number of times TCP will attempt to retransmit a packet on an established connection normally, without the extra effort of getting the network layers involved. Once we exceed this number of retransmits, we first have the network layer update the route if possible before each new retransmit. The default is the RFC specified minimum of 3.
tcp_retries2 (integer; **default: 15**; since Linux 2.2)
The maximum number of times a TCP packet is retransmitted in established state before giving up. The default value is 15, which corresponds to a duration of approximately between 13 to 30 minutes, depending on the retransmission timeout. The RFC 1122 specified minimum limit of 100 seconds is typically deemed too short.

**Comment by ghost** [comment]
这篇很棒，有定量和定性！
有两个疑问：
1）关于重传的问题“如上图所示，接收方已经收到了序号为 2-5 的数据，但是由于 TCP ACK 的语义是当前数据段前的全部数据段都已经被接收和处理，所以接收方无法发送 ACK 消息，由于发送方没有收到 ACK”

- 这里Jacobson在设计tcp的时候不是也设计了fast-retransmission机制么，这里应该会连续多次ack segment1？

2）3次握手的TCP开销是按header最大长度40算的么，但是正常握手的header不是20么？

**Comment by ghost** [comment]
问题2忽略哈

**Comment by draveness** [comment]
> 这篇很棒，有定量和定性！
> 有两个疑问：
> 1）关于重传的问题“如上图所示，接收方已经收到了序号为 2-5 的数据，但是由于 TCP ACK 的语义是当前数据段前的全部数据段都已经被接收和处理，所以接收方无法发送 ACK 消息，由于发送方没有收到 ACK”
> 
> * 这里Jacobson在设计tcp的时候不是也设计了fast-retransmission机制么，这里应该会连续多次ack segment1？
>   2）3次握手的TCP开销是按header最大长度40算的么，但是正常握手的header不是20么？

没太明白第一个问题，这里说的是接收方没有收到序号 1 的包，只收到 2-5 号

**Comment by ghost** [comment]
对的啊，收到了乱序数据，1没收到，2-5收到了；
这个时候jacobson的fast-restranmission algorithm和fast-recovery算法就是为了解决这个问题吧，
这个时候多次反复发送ack 1就可以了，tcp接收方会认为多次收到同一个ack seq会快速重新发送这个seq对应的segment。

RFC 2581 <https://tools.ietf.org/html/rfc2581>                 TCP Congestion Control               April 1999

3.2 <https://tools.ietf.org/html/rfc2581#section-3.2> Fast Retransmit/Fast Recovery

A TCP receiver SHOULD send an immediate duplicate ACK when an out-of-order segment arrives. <https://tools.ietf.org/html/rfc2581#page-6>

**Comment by draveness** [comment]
> 对的啊，收到了乱序数据，1没收到，2-5收到了； 这个时候jacobson的fast-restranmission algorithm和fast-recovery算法就是为了解决这个问题吧， 这个时候多次反复发送ack 1就可以了，tcp接收方会认为多次收到同一个ack seq会快速重新发送这个seq对应的segment。 RFC 2581 <https://tools.ietf.org/html/rfc2581> TCP Congestion Control April 1999 3.2 <https://tools.ietf.org/html/rfc2581#section-3.2> Fast Retransmit/Fast Recovery A TCP receiver SHOULD send an immediate duplicate ACK when an out-of-order segment arrives. <https://tools.ietf.org/html/rfc2581#page-6>

可以的，这个我之前真不清楚，我来加上补充的说明，非常感谢。

**Comment by renyuan431** [comment]
楼主您好，麻烦请教个问题，在文中您说“RFC2581 中提出的快速重传就会缓解上述问题，当接收方收到乱序的数据段时，会立刻发送两个连续的 ACK 触发发送方的重传。”，我看RFC上写的是3个ACK？是我理解错了吗？
“The fastretransmit algorithm uses the arrival of 3 duplicate ACKs (4identical ACKs without the arrival of any other intervening packets)as an indication that a segment has been lost.  After receiving 3duplicate ACKs, TCP performs a retransmission of what appears to bethe missing segment, without waiting for the retransmission timer toexpire.”

**Comment by draveness** [comment]
> 楼主您好，麻烦请教个问题，在文中您说“RFC2581 中提出的快速重传就会缓解上述问题，当接收方收到乱序的数据段时，会立刻发送两个连续的 ACK 触发发送方的重传。”，我看RFC上写的是3个ACK？是我理解错了吗？
> “The fastretransmit algorithm uses the arrival of 3 duplicate ACKs (4identical ACKs without the arrival of any other intervening packets)as an indication that a segment has been lost. After receiving 3duplicate ACKs, TCP performs a retransmission of what appears to bethe missing segment, without waiting for the retransmission timer toexpire.”

应该是这里写错了，我来改一下。不过这里有点歧义，没有说收到的数据段是不是发送了 ACK

**Comment by Aceralon** [comment]
这里的slow start的配图是否写错了呢，slow start阶段应该是每接收到一个ack就把当前流量加倍而不是加1。

**Comment by draveness** [comment]
> @Aceralon
> 这里的slow start的配图是否写错了呢，slow start阶段应该是每接收到一个ack就把当前流量加倍而不是加1。

它是每一个 RTT 加倍，每收到一个 ACK 只会加 1 https://en.wikipedia.org/wiki/TCP_congestion_control

**Comment by ZhouFyk** [comment]
> TCP 的 ACK 机制可能会导致发送方重新传输接收方已经收到了数据段

----
是否应该是「TCP 的 ACK 机制可能会导致发送方重新传输接收方已经收到的数据段」（「了」->「的」，或者也可以是「了的」）


---

*2021-03-09 UPDATES: 已修复*

**Comment by lxdlam** [comment]
> @draveness
> > @Aceralon
> > 这里的slow start的配图是否写错了呢，slow start阶段应该是每接收到一个ack就把当前流量加倍而不是加1。
> 
> 它是每一个 RTT 加倍，没收到一个 ACK 只会加 1 https://en.wikipedia.org/wiki/TCP_congestion_control

原文的的描述感觉存在歧义，应该同步修改为是“每收到一个 ACK 拥塞窗口大小加一”而不是每个“每经过一个 RTT 拥塞窗口大小加一”，或者“每个包在每个 RTT 内会让拥塞窗口大小加一”？

---

*2021-04-14 UPDATES: 已修复*

**Comment by DeloresC** [comment]
> RFC2581 中提出的快速重传就会缓解上述问题，当接收方收到乱序的数据段时，会立刻发送三个连续的 ACK 触发发送方的重传。

这里表达不太准确，并不会立刻发送三个连续的ack吧，比如有1，2，3，4表示包序号，1未收到，那么收到2时返回的也是ack1，3和4也是，这样才凑成了3个连续相同的ack，然后发送方再立刻重发1即可。

**Comment by superhqh** [comment]
计算通信双方窗口大小的公式里分子中的RTT是不是应该换成RTT/2

**Comment by aaa1999** [comment]
”所以接收方无法发送 ACK 消息“请问这句话是什么意思啊，没太懂

**Comment by bswaterb** [comment]
@aaa1999 
> ”所以接收方无法发送 ACK 消息“请问这句话是什么意思啊，没太懂

发送方本次发送了序列号为 1～5 的包，假设每个包代表一个 seq_number，如果接收方此时只接收到了 2～5 的包，因为缺了一个 1 号包没被收到，所以不能向发送方传输 [ACK seq = 6] 的包，也可能会向发送方持续传输[ACK seq = 1] 的包。对于发送方而言，太久没收到接收方的 ACK 响应，就需要对发送的消息进行重传，在文中举例的设计（类似于GBN回退N帧）下，发送方是不知道接收方已经接收到了本次发送的哪些包的，因此只能全部重传 1～5 号包。后续的 TCP 支持了选择重传（SACK），使得发送方能够得知接收方已经接收到了哪些包，因此只需要重传接收方没收到的包就行。
